### ISSUE TITLE
Pinning regression

### ISSUE THREAD

**Issue by o100ja** [senecajs/seneca#355]
The bump from 1.0 to 1.1 and 1.2 created a regression regarding the usage of ```seneca.pin``` method.

The problem is that on seneca >=1.1 the ```pin``` method returns an empty delegate collection until the **pin** event is triggered, which is emitted right after the **ready** event is emitted, which is in some cases too late.  Let me elaborate:
Let's say we have these plugins with the following methods stated in the order of being "required" by the ```seneca.use``` method: 
* pluginA
 * initA
 * msA1
 * msA2
* pluginB
 * initB
 * msB1
* pluginC
 * initC
 * msC1

And further more let's say that ```initB``` is dependent on ```msA1``` and ```msA2``` being called by ```initA``` where both of these have been "pinned" in the ```pluginB``` definition scope or in the ```initB``` scope.

With the current implementation this will fail since the delegate collection returned by ```seneca.pin``` is empty at this stage until the whole of seneca is "ready" which will occur only when ```pluginC``` is initialized and ready, which frankly has nothing to do with the either ```pluginA``` or ```pluginB```.

As you can see this is a tricky problem since the whole idea with having a **pin** event is to guarantee that the services are there and ready to be used, and the simplest way to do this was to wait for everything to be registered.

I see a couple of quick solutions but I don't like any of them:
1. Parametrize the **pin** event per plugin and emit the event when the plugin init action is completed.
2. ```seneca.pin``` should return a promise and should only be used in the plugin definition xor plugin init scope. 
3. Deprecate it completely, since it can lead to bigger problems when clustering and such. But this time inform the community about it 

We still may have an issue with the ```seneca.use``` method since this is where the problem truly lies.

In ANY case, we should inform the community that there is a problem with the concept of **pinning** and do it soon.

On Deprecation
--------------
Pinning is a dangerous idea that can be very useful for shortening the dev time but the problems are huge. I'm not sure that the current ```seneca.pin``` interface isn't a bit misleading, especially since there is no mention in the documentation and tutorials that it may be deprecated. 
Personally I like the concept but it MUST be restricted to certain use cases thus reducing the risk of possible issues with clustering and such concepts.

As a example, I've been using ```seneca.pin``` to create delegates to certain Data Access Services that are wrappers around Knex functionalities, which are encapsulated in a separate plugin (```pluginA```) and being used the the Business Logic Services (```pluginB```) locally for now. This is where having the possibility to restrict certain actions or plugins to being local only would come in hand, since then we could **pin** only actions that are declared as being local only.

**Comment by rjrodger** [comment]
Agreed - pinning is a dead-end feature that has proven to have too much downside in practice, and is due for deprecation this year.

We will however work on fixing this regression! :)

**Comment by mcdonnelldean** [comment]
@o100ja On the community contact. I just wanted to say this is something we are going to get better at quickly. We are putting plans in place to make this much easier (The first of which was the move to the org). All of the team really want this so we are working through what is practical.

I just wanted to be clear we know this is a pain point and we are moving to improve it.

**Comment by geek** [comment]
@o100ja how is this for a potential workaround? If this isn't what you are fully describing, can you please update the example below:

```js
    var seneca = Seneca({ log: 'silent' })

    var pluginA = function () {
      this.add({ init: 'pluginA' }, function (msg, cb) {
        process.nextTick(cb)
      })

      this.add({ role: 'pluginA', cmd: 'msA1' }, function (msg, cb) {
        cb(null, { result: 'msA1' })
      })

      this.add({ role: 'pluginA', cmd: 'msA2' }, function (msg, cb) {
        cb(null, { result: 'msA2' })
      })

      return {
        name: 'pluginA'
      }
    }

    var pluginB = function () {
      this.add({ init: 'pluginB' }, function (msg, cb) {
        this.once('pin', function () {
          var api = this.pin({ role: 'pluginA', cmd: '*' })
          api.msA1({ msg: 'hi' }, function (err, message) {
            expect(err).to.not.exist()
            expect(message.result).to.equal('msa1')
          })
        })
        cb()
      })

      this.add({ cmd: 'msB1' }, function (msg, cb) {
        cb(null, { result: 'msB1' })
      })

      return {
        name: 'pluginB'
      }
    }

    var pluginC = function () {
      this.add({ init: 'pluginC' }, function (msg, cb) {
        process.nextTick(cb)
      })

      this.add({ cmd: 'msC1' }, function (msg, cb) {
        cb(null, { result: 'msC1' })
      })

      return {
        name: 'pluginC'
      }
    }

    seneca.use(pluginA)
    seneca.use(pluginB)
    seneca.use(pluginC)
    seneca.ready(function () {
      done()
    })
```

**Comment by o100ja** [comment]
Tnx @geek but I already tried similar solutions and it simply doesn't work for me at this point.
Regardless since were definitely considering deprecation, I have a proposal for a quick solution that works for me and it is not expensive:

Introduce a new option (similar to timeout) that forces the pin process to be synchronous. Not sure about the parameter name (synchronous_pin: true|false) or something similar. 
This parameter can be global(```so```)  or local (```pinopts```) or both
Then all we need to do is to change the ```make_pin``` function to include this parameter in the condition on line 553, similar to something like:
```
if (private$._isReady || so.synchronous_pin || pinopts.synchronous) { // <- HERE's the change
  calcPin() 
} else {
  root.once('pin', calcPin)
}
```
By default this parameter will be false (or falsy in any case), which ensures that the new "default" pinning behaviour will not be changed. Also we're transferring the "responsibility" of the pinning behaviour to the developer. If he/she knows what they are doing (basically has an understanding of in which order the ```seneca.use``` method should be invoked) they will not be encumbered by this regression.

I've already tested this on my project and it didn't present any problems.

Naturally this only makes sense as long as we agree that pinning **WILL** be deprecated in the future, and it is thoroughly documented, starting with the announcement that the ```pin``` method will be deprecated and that the current default behaviour has changed but can be reverted using the options parameter (sync_pin).  While on the subject of docs, thanks @mcdonnelldean for the good news.
